lib/linkerd2-proxy-transport/tests/tls_accept.rs

```diff
@@ -0,0 +1,320 @@
+        |conn| write_then_read(conn, PING),
+        Conditional::None(tls::ReasonForNoIdentity::Disabled),
+        |conn| read_then_write(conn, PING.len(), PONG),
+    );
+    assert_eq!(client_result.is_tls(), false);
+    assert_eq!(&client_result.result.expect("pong")[..], PONG);
+    assert_eq!(server_result.is_tls(), false);
+    assert_eq!(&server_result.result.expect("ping")[..], PING);
+}
+
+#[test]
+fn proxy_to_proxy_tls_works() {
+    let server_tls = test_util::FOO_NS1.validate().unwrap();
+    let client_tls = test_util::BAR_NS1.validate().unwrap();
+    let (client_result, server_result) = run_test(
+        Conditional::Some((client_tls, server_tls.tls_server_name())),
+        |conn| write_then_read(conn, PING),
+        Conditional::Some(server_tls),
+        |conn| read_then_write(conn, PING.len(), PONG),
+    );
+    assert_eq!(client_result.is_tls(), true);
+    assert_eq!(&client_result.result.expect("pong")[..], PONG);
+    assert_eq!(server_result.is_tls(), true);
+    assert_eq!(&server_result.result.expect("ping")[..], PING);
+}
+
+#[test]
+fn proxy_to_proxy_tls_pass_through_when_identity_does_not_match() {
+    let server_tls = test_util::FOO_NS1.validate().unwrap();
+
+    // Misuse the client's identity instead of the server's identity. Any
+    // identity other than `server_tls.server_identity` would work.
+    let client_tls = test_util::BAR_NS1.validate().expect("valid client cert");
+    let client_target = test_util::BAR_NS1.crt().name().clone();
+
+    let (client_result, server_result) = run_test(
+        Conditional::Some((client_tls, client_target)),
+        |conn| write_then_read(conn, PING),
+        Conditional::Some(server_tls),
+        |conn| read_then_write(conn, START_OF_TLS.len(), PONG),
+    );
+
+    // The server's connection will succeed with the TLS client hello passed
+    // through, because the SNI doesn't match its identity.
+    assert_eq!(client_result.is_tls(), false);
+    assert!(client_result.result.is_err());
+    assert_eq!(server_result.is_tls(), false);
+    assert_eq!(&server_result.result.unwrap()[..], START_OF_TLS);
+}
+
+struct Transported<R> {
+    /// The value of `Connection::peer_identity()` for the established connection.
+    ///
+    /// This will be `None` if we never even get a `Connection`.
+    peer_identity: Option<tls::PeerIdentity>,
+
+    /// The connection's result.
+    result: Result<R, io::Error>,
+}
+
+impl<R> Transported<R> {
+    fn is_tls(&self) -> bool {
+        self.peer_identity
+            .as_ref()
+            .map(|i| i.is_some())
+            .unwrap_or(false)
+    }
+}
+
+/// Runs a test for a single TCP connection. `client` processes the connection
+/// on the client side and `server` processes the connection on the server
+/// side.
+fn run_test<C, CF, CR, S, SF, SR>(
+    client_tls: tls::Conditional<(CrtKey, Name)>,
+    client: C,
+    server_tls: tls::Conditional<CrtKey>,
+    server: S,
```

Comment:
TIOLI: IMO passing this many params to this function makes it a little hard to tell what's going on when looking at the tests...WDYT about combining them into `Server` and `Client` structs or something, so that the purpose of the params is clearer at the callsite?

Not a big deal in any case.